src/libopensc/pkcs15-sc-hsm.c

```diff
@@ -1028,10 +1028,16 @@ static int sc_pkcs15emu_sc_hsm_add_pubkey(sc_pkcs15_card_t *p15card, u8 *efbin,
 	memset(&pubkey_obj, 0, sizeof(pubkey_obj));
 
 	r = sc_pkcs15_encode_pubkey(ctx, &pubkey, &pubkey_obj.content.value, &pubkey_obj.content.len);
+	if (r != SC_SUCCESS)
+		sc_pkcs15_erase_pubkey(&pubkey);
 	LOG_TEST_RET(ctx, r, "Could not encode public key");
 	r = sc_pkcs15_encode_pubkey(ctx, &pubkey, &pubkey_info.direct.raw.value, &pubkey_info.direct.raw.len);
+	if (r != SC_SUCCESS)
+		sc_pkcs15_erase_pubkey(&pubkey);
 	LOG_TEST_RET(ctx, r, "Could not encode public key");
 	r = sc_pkcs15_encode_pubkey_as_spki(ctx, &pubkey, &pubkey_info.direct.spki.value, &pubkey_info.direct.spki.len);
+	if (r != SC_SUCCESS)
+		sc_pkcs15_erase_pubkey(&pubkey);
 	LOG_TEST_RET(ctx, r, "Could not encode public key");
```

Comment:
I would suggest to keep both in the condition to be clear on the first sight both of the commands are error handling and they return.
```suggestion
	if (r != SC_SUCCESS) {
		sc_pkcs15_erase_pubkey(&pubkey);
		LOG_TEST_RET(ctx, r, "Could not encode public key");
	}
	r = sc_pkcs15_encode_pubkey(ctx, &pubkey, &pubkey_info.direct.raw.value, &pubkey_info.direct.raw.len);
	if (r != SC_SUCCESS) {
		sc_pkcs15_erase_pubkey(&pubkey);
		LOG_TEST_RET(ctx, r, "Could not encode public key");
	}
	r = sc_pkcs15_encode_pubkey_as_spki(ctx, &pubkey, &pubkey_info.direct.spki.value, &pubkey_info.direct.spki.len);
	if (r != SC_SUCCESS) {
		sc_pkcs15_erase_pubkey(&pubkey);
		LOG_TEST_RET(ctx, r, "Could not encode public key");
	}
```